### PR TITLE
Fix bugs in `insert_and_find`

### PR DESCRIPTION
**Pull request by PointKernel**
This PR fixes bugs in the new `insert_and_find` implementation where the function should not return before the payload is updated. It also migrates the related tests to test the new map.

### CODE REVIEW

**Comment by sleeepyjack** [comment]
Gnaaaahhh GH mobile UI tricked me again...I accidentally marked this as ready for review and now I can't undo that on mobile. 🫠

**Comment by sleeepyjack** [comment]
Proposed spin wait approach: https://godbolt.org/z/hfGxoh9h8

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh [hunk 1]
```diff
@@ -397,17 +397,44 @@ class open_addressing_ref_impl {
         auto* window_ptr  = (storage_ref_.data() + *probing_iter)->data();
 
         // If the key is already in the container, return false
-        if (eq_res == detail::equal_result::EQUAL) { return {iterator{&window_ptr[i]}, false}; }
+        if (eq_res == detail::equal_result::EQUAL) {
+          if constexpr (sizeof(value_type) > 8) {
+            using mapped_type           = decltype(this->empty_slot_sentinel_.second);
+            auto const expected_payload = this->empty_slot_sentinel_.second;
+            auto ref =
+              cuda::atomic<mapped_type, cuda::thread_scope_device>{(window_ptr + i)->second};
```
Comment (sleeepyjack) [inline include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh]: `cuda::atomic_ref`, not `cuda::atomic`

Comment (PointKernel) [reply include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh]: bug of this year.

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh [hunk 2]
```diff
@@ -385,6 +385,12 @@ class open_addressing_ref_impl {
   __device__ thrust::pair<iterator, bool> insert_and_find(Value const& value) noexcept
   {
     static_assert(cg_size == 1, "Non-CG operation is incompatible with the current probing scheme");
+#if __CUDA_ARCH__ < 700
+    // Spinning to ensure that the write to the value part took place requires
+    // independent thread scheduling introduced with the Volta architecture.
+    static_assert(cuco::detail::is_packable<value_type>(),
+                  "insert_and_find is not supported for unpackable data on pre-Volta GPUs.");
```
Comment (sleeepyjack) [inline include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh]:
```suggestion
                  "insert_and_find is not supported for pair types larger than 8 bytes on pre-Volta GPUs.");
```

"packable data" sounds misleading.

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh [hunk 3]
```diff
@@ -397,21 +403,38 @@ class open_addressing_ref_impl {
...
         if (eq_res == detail::equal_result::EMPTY or
             cuco::detail::bitwise_compare(this->extract_key(window_slots[i]),
                                           this->erased_key_sentinel())) {
-          switch ([&]() {
+          auto const res = [&]() {
             if constexpr (sizeof(value_type) <= 8) {
               return packed_cas(window_ptr + i, window_slots[i], value);
             } else {
-              return cas_dependent_write(window_ptr + i, window_slots[i], value);
+              auto const expected_payload = this->empty_slot_sentinel_.second;
```
Comment (sleeepyjack) [inline include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh]:
```suggestion
```

This is never used.

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh [hunk 4]
```diff
@@ -397,21 +403,38 @@ class open_addressing_ref_impl {
...
+              auto const res = cas_dependent_write(window_ptr + i, window_slots[i], value);
+              return res;
```
Comment (sleeepyjack) [inline include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh]:
```suggestion
              return cas_dependent_write(window_ptr + i, window_slots[i], value);
```

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh [hunk 5]
```diff
@@ -441,6 +464,13 @@ class open_addressing_ref_impl {
   __device__ thrust::pair<iterator, bool> insert_and_find(
     cooperative_groups::thread_block_tile<cg_size> const& group, Value const& value) noexcept
   {
+#if __CUDA_ARCH__ < 700
+    // Spinning to ensure that the write to the value part took place requires
+    // independent thread scheduling introduced with the Volta architecture.
```
Comment (sleeepyjack) [inline include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh:468]: Good catch! That wasn't on my radar.

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh [hunk 6]
```diff
@@ -475,6 +505,12 @@ class open_addressing_ref_impl {
       if (group_finds_equal) {
         auto const src_lane = __ffs(group_finds_equal) - 1;
         auto const res      = group.shfl(reinterpret_cast<intptr_t>(slot_ptr), src_lane);
+        if (group.thread_rank() == src_lane) {
```
Comment (sleeepyjack) [inline include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh:508]: I think all threads in the group need to spin wait on the payload. Or alternatively one rank waits and then a subsequent call to `group.sync()`.

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh [hunk 7]
```diff
@@ -1000,7 +1048,8 @@ class open_addressing_ref_impl {
   {
     using mapped_type = decltype(this->empty_slot_sentinel_.second);
 
-    auto const expected_key = expected.first;
+    auto const expected_key     = expected.first;
+    auto const expected_payload = expected.second;
```
Comment (sleeepyjack) [inline include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh]:
```suggestion
```
I might be blind but I don't see this variable being used anywhere.

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh [hunk 8]
```diff
@@ -1054,6 +1104,27 @@ class open_addressing_ref_impl {
...
+   *
+   * @param slot The target slot to check payload with
+   * @param sentinel The slot sentinel value
+   */
+  template <typename T>
+  __device__ void wait_for_payload(T& slot, T const& sentinel) const noexcept
+  {
+    auto ref = cuda::atomic_ref<T, Scope>{slot};
+    T current;
+    // TODO exponential backoff strategy
```
Comment (sleeepyjack) [inline include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh:1124]: On my TODO list

**Review by sleeepyjack — CHANGES_REQUESTED** [review]

File: include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh (hunk 6 above)
Comment (PointKernel) [reply include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh:508]: `n` threads in a group call the spin will cause at least `n` sequential memory load which would be really bad for performance.

Comment (PointKernel) [reply include/cuco/detail/open_addressing/open_addressing_ref_impl.cuh:508]: but good point on the group sync

**Comment by PointKernel** [comment]
How come the changes unrelated to `erase` failed an erase test? :sob:

**Comment by sleeepyjack** [comment]
> How come the changes unrelated to `erase` failed an erase test? 😭

I've seen the `erase` test spuriously fail in the past. Seems flaky… which means there might be a hidden race condition or something akin.

**Comment by PointKernel** [comment]
> > How come the changes unrelated to `erase` failed an erase test? 😭
> 
> I've seen the `erase` test spuriously fail in the past. Seems flaky… which means there might be a hidden race condition or something akin.

Looking into this. Would be a separate PR though.

**Review by sleeepyjack — APPROVED** [review]

**Pull request merged**